### PR TITLE
dhcp start/stop made compatible for multiple interfaces

### PR DESCRIPTION
**Pull request by vlotech**
The net_dhcpv4_stop() function stops the event listener for any IF_UP events previously. In case multiple interfaces are used and optionally being switched over, this could result in disabling dhcp unintentionally. 

Signed-off-by: Vincent van der Locht <vincent@vlotech.nl>

### CODE REVIEW

**Review by tbursztyka — CHANGES_REQUESTED** [review]

**Comment by vlotech** [comment]
Based on @tbursztyka comment I've updated the pull-request. 
The callback is initialized at the init function and added/removed in the start/stop function based on the interface list being empty or not (first added, last removed)
The event handler checks if the interface is in the list before acting on it.

File: subsys/net/ip/dhcpv4.c [hunk 1]
```diff
@@ -1090,23 +1090,15 @@ void net_dhcpv4_start(struct net_if *iface)
...
 
 void net_dhcpv4_stop(struct net_if *iface)
 {
-	net_mgmt_del_event_callback(&mgmt4_cb);
-
 	switch (iface->config.dhcpv4.state) {
 	case NET_DHCPV4_DISABLED:
 		break;
 
+	case NET_DHCPV4_RENEWING:
```
[the remark this answers is not in the document]
Comment (vlotech) [reply subsys/net/ip/dhcpv4.c]: I'll make a separate patch for this

File: subsys/net/ip/dhcpv4.c [hunk 2]
```diff
@@ -991,6 +992,17 @@ static enum net_verdict net_dhcpv4_input(struct net_conn *conn,
 static void dhcpv4_iface_event_handler(struct net_mgmt_event_callback *cb,
 				       u32_t mgmt_event, struct net_if *iface)
 {
+	sys_snode_t *node = NULL;
+
+	SYS_SLIST_FOR_EACH_NODE(&dhcpv4_ifaces, node) {
+		if (node == &iface->config.dhcpv4.node) {
+		  break;
```
Comment (tbursztyka) [inline subsys/net/ip/dhcpv4.c]: and your indentation is wrong: tabs only (plus spaces when it comes to parameters alignment) please.

File: subsys/net/ip/dhcpv4.c [hunk 3]
```diff
@@ -991,6 +992,17 @@ static enum net_verdict net_dhcpv4_input(struct net_conn *conn,
...
+	  }
+	}
+
+	if (node == NULL)
+		 return;
```
Comment (tbursztyka) [inline subsys/net/ip/dhcpv4.c]: {} even for one liner statement

**Review by tbursztyka — CHANGES_REQUESTED** [review]

**Comment by vlotech** [comment]
@tbursztyka @jukkar Would it be possible to take a look at this pull request. Based on the given comments, I hopefully updated the commit. If not, please let me know.

**Comment by pfalcon** [comment]
My 2 cents: I find the commit description here unclear:

> The net_dhcpv4_stop() function stops the event listener for any IF_UP events previously.

I may imagine a word (participle) is missing here, or perhaps I don't understand the meaning of this sentence completely.

> In case multiple interfaces are used and optionally being switched over, this could result in disabling dhcp unintentionally.

I don't understand what "optionally being switched over" means here. The usages of "switch over" I know are also usually followed with "to"/"from".


My point is that I couldn't match the description of the patch with the changes included in it (even if I understand the DHCP code well, which I don't). So, relying on @jukkar/@tbursztyka for review.

**Review by tbursztyka — APPROVED** [review]

**Review by jukkar — APPROVED** [review]

**Pull request merged**